velox/exec/tests/utils/ExpressionBuilder.h

```diff
@@ -0,0 +1,387 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/parse/Expressions.h"
+
+namespace facebook::velox::expr_builder {
+
+/// Fluent Expression Builder.
+///
+/// This file contains fluent methods that make it convenient to create
+/// (untyped) expression trees. This provides similar functionality to a SQL
+/// parser, without briging dependency on external libraries or bringing leaked
+/// semantics from other systems.
+///
+/// The untyped expressions can then be turned into typed expressions ready for
+/// execution using type binding from `core::Expressions::inferTypes()`.
+///
+/// The API provided is as close to the actual expression trees as possible.
+/// Filters, arithmetics, function calls, literals, aliases, and more are
+/// supported with this API.
+///
+/// For example, to create a column reference, you can:
+///
+/// > using namespace expr_builder;
+/// > ExprPtr e = col("c0");
+///
+/// You can also use the "_c" C++ literal provided:
+///
+/// > ExprPtr e = "c0"_c;
+///
+/// Nested column references can be specified as either:
+///
+/// > ExprPtr e = col("parent", "child");
+/// > ExprPtr e = col("parent").subfield("child");
+///
+/// To debug the expression generated, you can simply:
+///
+/// > LOG(INFO) << *e;
+///
+/// Comparisons and other expressions can be fluently created using C++
+/// overloaded operators:
+///
+/// > ExprPtr e = col("c") > 10; // "c > 10"
+/// > ExprPtr e = col("c") != "bar"; // "c != 'bar'"
+/// > ExprPtr e = col("c") == nullptr; // "c = null"
+///
+/// C++ literals are automatically converted into ConstantExpr (expression
+/// literals) when part of an expression. To explicitly create a literal you can
+/// use:
+///
+/// > ExprPtr e = lit(10.3);
+/// > ExprPtr e = lit("str");
+///
+/// Casts can be done using one of the two formats:
+///
+/// > ExprPtr e = lit(3).cast(TINYINT());
+/// > ExprPtr e = cast("str", VARBINARY());
+///
+/// Null checking filters:
+///
+/// > ExprPtr e = isNull(col("c")) // "c is null"
+/// > ExprPtr e = !isNull(col("c")) // "c is not null"
+///
+/// Conjuncts and "between":
+///
+/// > ExprPtr e = col("a") && col("b") || col("c"); // "a AND b OR c"
```

Comment:
this feel error prone; perhaps, use explicit parentheses to clarify the intended order of operations

Reply:
Not really; it will follow operator precedence rules in C++, which are deterministic and well defined. I can add parenthesis  here for clarity though.